sympy/functions/combinatorial/factorials.py

```diff
@@ -331,7 +331,7 @@ def eval(cls, arg):
 
         if arg.is_Number:
             if arg.is_infinite:
-                return
+                raise ValueError("argument must be nonnegative integer or odd integer")
```

Comment:
If the line above were replaced by ``if not arg.is_Integer``, then it would catch the infinities and, in addition, arg would be an Integer henceforth.